test/unit/test_recipe.cpp

```diff
@@ -46,18 +46,23 @@ namespace {
             return cell_kind::cable;
         }
         std::vector<gap_junction_connection> gap_junctions_on(cell_gid_type gid) const override {
+            if (gid >= gap_junctions_.size()) return {};
```

Comment:
Here and below, if this is to keep the sanitizer happy, I'd go with `gap_junctions_.at(gid)` instead. Then at least in a non-sanitized environment, we'd get a useful exception if the implicit contract is broken.

Reply:
The test assumes a `arb::bad_source/_target` to be thrown; this change would break that assumption, right?

Reply:
Yes: a bad source error _should_ be thrown, but isn't because of a bug in communication.cpp:
```
diff --git a/arbor/communication/communicator.cpp b/arbor/communication/communicator.cpp
index b93638f8..16b08fb5 100644
--- a/arbor/communication/communicator.cpp
+++ b/arbor/communication/communicator.cpp
@@ -80,11 +80,10 @@ communicator::communicator(const recipe& rec,
     for (const auto& cell: gid_infos) {
         auto num_targets = rec.num_targets(cell.gid);
         for (auto c: cell.conns) {
-            auto num_sources = rec.num_sources(c.source.gid);
             if (c.source.gid >= num_total_cells) {
                 throw arb::bad_connection_source_gid(cell.gid, c.source.gid, num_total_cells);
             }
-            if (c.source.index >= num_sources) {
+            if (auto num_sources = rec.num_sources(c.source.gid); c.source.index >= num_sources) {
                 throw arb::bad_connection_source_lid(cell.gid, c.source.index, num_sources);
             }
             if (c.dest >= num_targets) {
```